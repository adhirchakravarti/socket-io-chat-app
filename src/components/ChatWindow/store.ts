import {createStore, combineReducers, applyMiddleware} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import createSocketIoMiddleware from "redux-socket.io";
import {SERVER_URL} from "./constants";
import chatReducer from "./reducer";
import io from "socket.io-client";

const socket = io(SERVER_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, "chatServer/");
const middleWares = [socketIoMiddleware];
// import {initialState} from "./reducer";

// function reducer(state = {}, action) {
//     switch (action.type) {
//         case "message":
//             return Object.assign({}, {message: action.data});
//         default:
//             return state;
//     }
// }

const rootReducer = combineReducers({
    chatReducer
});
const enhancer = composeWithDevTools(applyMiddleware(...middleWares));

const store = createStore(rootReducer, enhancer);

store.subscribe(() => {
    console.log("new client state", store.getState());
});
// store.dispatch({type: "chatServer/hello", data: "Hello!"});
setTimeout(() => {
    store.dispatch({type: "chatServer/SET_USERNAME", payload: {userName: "Jack"}});
}, 20000);

export default store;
