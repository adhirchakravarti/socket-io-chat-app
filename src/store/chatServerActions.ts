import {Message, Settings} from "../types/index";
import {
    SEND_MESSAGE,
    SET_USERNAME,
    RECEIVE_MESSAGE,
    GET_INITIAL_USERNAME,
    SET_THEME,
    CHANGE_SETTINGS,
    RESET_SETTINGS,
    UNREAD_MESSAGE,
    RESET_UNREAD_MESSAGE_COUNT
} from "./chatServerConstants";

interface SendMessageAction {
    type: typeof SEND_MESSAGE;
    payload: {
        message: Message;
    };
}

interface ReceiveMessageAction {
    type: typeof RECEIVE_MESSAGE;
    payload: {
        message: Message;
    };
}

interface SetUsernameAction {
    type: typeof SET_USERNAME;
    payload: {
        userName: string;
    };
}

interface GetInitialUserNameAction {
    type: typeof GET_INITIAL_USERNAME;
}

interface SetThemeAction {
    type: typeof SET_THEME;
    payload: {
        theme: string;
    };
}

interface ChangeSettingsAction {
    type: typeof CHANGE_SETTINGS;
    payload: {
        settings: Settings;
    };
}

interface ResetSettingsAction {
    type: typeof RESET_SETTINGS;
}

interface UnreadMessageAction {
    type: typeof UNREAD_MESSAGE;
}

interface ResetUnreadMessageCountAction {
    type: typeof RESET_UNREAD_MESSAGE_COUNT;
}

export function sendMessage(message: Message): SendMessageAction {
    return {
        type: SEND_MESSAGE,
        payload: {
            message
        }
    };
}

export function receiveMessage(message: Message): ReceiveMessageAction {
    return {
        type: RECEIVE_MESSAGE,
        payload: {
            message
        }
    };
}

export function setUsername(userName: string): SetUsernameAction {
    return {
        type: SET_USERNAME,
        payload: {
            userName
        }
    };
}

export function getInitialUserName(): GetInitialUserNameAction {
    return {
        type: GET_INITIAL_USERNAME
    };
}

export function setTheme(theme: string): SetThemeAction {
    return {
        type: SET_THEME,
        payload: {
            theme
        }
    };
}

export function changeSettings(settings: Settings): ChangeSettingsAction {
    return {
        type: CHANGE_SETTINGS,
        payload: {
            settings
        }
    };
}

export function resetSettings(): ResetSettingsAction {
    return {
        type: RESET_SETTINGS
    };
}

export function unreadMessage(): UnreadMessageAction {
    return {
        type: UNREAD_MESSAGE
    };
}

export function resetUnreadMessageCount(): ResetUnreadMessageCountAction {
    return {
        type: RESET_UNREAD_MESSAGE_COUNT
    };
}

export type ChatServerActionTypes =
    | SendMessageAction
    | ReceiveMessageAction
    | SetUsernameAction
    | GetInitialUserNameAction
    | SetThemeAction
    | ChangeSettingsAction
    | ResetSettingsAction
    | UnreadMessageAction
    | ResetUnreadMessageCountAction;
