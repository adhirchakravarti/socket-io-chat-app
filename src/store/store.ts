import {createStore, combineReducers, applyMiddleware} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import createSocketIoMiddleware from "redux-socket.io";
import {SERVER_URL} from "./constants";
import chatReducer from "./reducer";
import io from "socket.io-client";

const socket = io(SERVER_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, "chatServer/");
const middleWares = [socketIoMiddleware];

const rootReducer = combineReducers({
    chatReducer
});
const enhancer = composeWithDevTools(applyMiddleware(...middleWares));

const store = createStore(rootReducer, enhancer);

const savedSettings = JSON.parse(sessionStorage.getItem("settings"));
if (savedSettings && Object.keys(savedSettings).length > 0) {
    const savedUserName = savedSettings.userName;
    store.dispatch({type: "chatClient/LOAD_SETTINGS"});
    setTimeout(
        () => store.dispatch({type: "chatServer/SET_USERNAME", payload: {userName: savedUserName}}),
        1000
    );
}

export default store;
