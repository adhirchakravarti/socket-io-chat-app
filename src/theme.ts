import {createMuiTheme} from "@material-ui/core";

export const lightTheme = createMuiTheme({
    palette: {
        type: "light",
        primary: {
            main: "#0d47a1"
        },
        secondary: {
            main: "#689f38"
        }
    }
});

export const darkTheme = createMuiTheme({
    palette: {
        type: "dark",
        primary: {
            main: "#29b6f6"
        },
        secondary: {
            main: "#689f38"
        },
        background: {
            paper: "#212121",
            default: "#000"
        },
        contrastThreshold: 5
    }
});
