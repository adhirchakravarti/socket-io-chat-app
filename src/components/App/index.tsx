import React from "react";
import {BrowserRouter as Router, Switch, Route, Redirect} from "react-router-dom";
import CssBaseline from "@material-ui/core/CssBaseline";
import Container from "@material-ui/core/Container";
import Grid from "@material-ui/core/Grid";
import GlobalResponsiveFont from "../../globalResponsiveFont";

import Header from "../Header";
import UserSettings from "../UserSettings";
import ChatWindow from "../ChatWindow";
import TabPanel from "../TabPanel";

const tabs = [
    {label: "Chat", link: "chat"},
    {label: "Settings", link: "settings"}
];

function App(): JSX.Element {
    return (
        <Router>
            <CssBaseline />
            <GlobalResponsiveFont />
            <Switch>
                <Container maxWidth="xl">
                    <Grid
                        container
                        direction="column"
                        justify="center"
                        alignItems="stretch"
                        alignContent="stretch"
                        xl={12}
                        spacing={2}
                    >
                        <Grid item xl={12} lg={12} md={12} sm={12} xs={12}>
                            <Header title="Docler Chat App" />
                            <TabPanel tabs={tabs} />
                        </Grid>
                        <Grid item xl={12} lg={12} md={12} sm={12} xs={12}>
                            <Route path="/" exact>
                                <Redirect to="/chat" />
                            </Route>
                            <Route path="/chat" component={ChatWindow} />
                            <Route path="/settings" component={UserSettings} />
                        </Grid>
                    </Grid>
                </Container>
            </Switch>
        </Router>
    );
}

export default App;
