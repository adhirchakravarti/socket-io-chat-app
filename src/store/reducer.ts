import {original, produce, setAutoFreeze} from "immer";
setAutoFreeze(false);

import {
    SET_USERNAME_SUCCESS,
    RECEIVE_MESSAGE,
    SOCKET_ID,
    CHANGE_SETTINGS,
    RESET_SETTINGS,
    UNREAD_MESSAGE,
    RESET_UNREAD_MESSAGE_COUNT
} from "./chatServerConstants";
import {ChatState, Message} from "../types/index";
import {ChatServerActionTypes} from "./chatServerActions";
import {Reducer} from "react";

export const initialState: ChatState = {
    messages: [],
    userName: "",
    socketId: "",
    theme: "light",
    clock: "24",
    sendMessageOnCtrlEnter: "false",
    unreadMessageCount: 0
};

const chatReducer: Reducer<ChatState, ChatServerActionTypes> = produce((draft, action) => {
    switch (action.type) {
        case SOCKET_ID: {
            const {socketId} = action.payload;
            draft.socketId = socketId;
            break;
        }

        case SET_USERNAME_SUCCESS: {
            const {userName} = action.payload;
            draft.userName = userName;
            break;
        }

        case RECEIVE_MESSAGE: {
            const {message} = action.payload;
            const {messages: existingMessages} = original(draft);
            const messageIndex = existingMessages.findIndex(
                (msg: Message) => msg.id === message.id
            );
            if (messageIndex < 0) {
                existingMessages.push(message);
            }
            draft.messages = [...existingMessages];
            break;
        }

        case CHANGE_SETTINGS: {
            const {
                settings: {userName, theme, clock, sendMessageOnCtrlEnter}
            } = action.payload;
            draft.userName = userName;
            draft.theme = theme;
            draft.clock = clock;
            draft.sendMessageOnCtrlEnter = sendMessageOnCtrlEnter;
            break;
        }

        case RESET_SETTINGS: {
            draft.userName = original(draft).userName;
            draft.theme = "light";
            draft.clock = "24";
            draft.sendMessageOnCtrlEnter = "false";
            break;
        }

        case UNREAD_MESSAGE: {
            let unreadMessageCount = original(draft).unreadMessageCount;
            unreadMessageCount += 1;
            draft.unreadMessageCount = unreadMessageCount;
            break;
        }

        case RESET_UNREAD_MESSAGE_COUNT: {
            draft.unreadMessageCount = 0;
            break;
        }
    }
}, initialState);

export default chatReducer;
