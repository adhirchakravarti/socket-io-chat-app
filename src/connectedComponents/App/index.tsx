import React, {useState, useEffect} from "react";
import {BrowserRouter as Router, Switch, Route, Redirect} from "react-router-dom";
import {connect} from "react-redux";
import {Dispatch} from "redux";

import Box from "@material-ui/core/Box";
import MailIcon from "@material-ui/icons/Mail";
import SettingsIcon from "@material-ui/icons/Settings";

import {ThemeProvider, makeStyles, createStyles} from "@material-ui/core/styles";
import GlobalResponsiveFont from "../../globalResponsiveFont";
import CssBaseline from "@material-ui/core/CssBaseline";
import useWindowHeight from "../../utils/useWindowHeight";

import Header from "../../components/Header";
import UserSettings from "../UserSettings";
import ChatWindow from "../ChatWindow";
import TabPanel from "../../components/TabPanel";
import BadgeIcon from "../../components/BadgeIcon";
import {lightTheme, darkTheme} from "../../theme";
import {useMediaQuery} from "react-responsive";
import {loadSettings} from "../../store/actions";

const useStyles = makeStyles(() =>
    createStyles({
        gridContainer: {
            maxHeight: (props: {height: number}) => {
                return props.height ? props.height : "100%";
            },
            display: "flex",
            flexDirection: "column",
            flexGrow: 1,
            margin: "0rem",
            "& > *:first-child": {
                marginBottom: "2rem"
            }
        },
        content: {
            maxHeight: "75vh"
        }
    })
);

const mapStateToProps = (state) => {
    return {
        socketId: state.chatReducer.socketId,
        theme: state.chatReducer.theme,
        unreadMessageCount: state.chatReducer.unreadMessageCount
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        loadSettingsAction: () => dispatch(loadSettings())
    };
};

interface AppProps {
    socketId: string;
    theme: string;
    unreadMessageCount: number;
    loadSettingsAction: () => Dispatch;
}

function App({
    socketId,
    theme,
    unreadMessageCount,
    loadSettingsAction
}: AppProps): React.FunctionComponentElement<AppProps> {
    const [visible, setVisible] = useState(true);
    const {height} = useWindowHeight();
    const classes = useStyles({height});
    const isPortrait = useMediaQuery({query: "(orientation: portrait)"});
    const makeVisible = () => {
        setVisible((prevState) => !prevState);
    };
    useEffect(() => {
        if (socketId) {
            const savedSettings = JSON.parse(sessionStorage.getItem("settings"));
            if (savedSettings && Object.keys(savedSettings).length > 0) {
                loadSettingsAction();
            }
        }
    }, [socketId]);
    useEffect(() => {
        if (unreadMessageCount > 0) {
            setVisible(false);
        } else {
            setVisible(true);
        }
    }, [unreadMessageCount]);
    useEffect(() => {
        let id;
        if (unreadMessageCount > 0) {
            id = setInterval(makeVisible, 700);
        } else {
            clearInterval(id);
            setVisible(true);
        }
        return () => clearInterval(id);
    }, [visible]);
    const tabs = [
        {
            label: "Chat",
            link: "chat",
            icon: (
                <BadgeIcon
                    content={unreadMessageCount}
                    childIcon={<MailIcon />}
                    visible={visible}
                />
            )
        },
        {label: "Settings", link: "settings", icon: <SettingsIcon />}
    ];

    return (
        <ThemeProvider theme={theme === "light" ? lightTheme : darkTheme}>
            <CssBaseline />
            <GlobalResponsiveFont />
            <Router>
                <Box className={classes.gridContainer}>
                    <Box>
                        <Header title="Chat-App" />
                        <TabPanel tabs={tabs} />
                    </Box>
                    <Box className={classes.content}>
                        <Switch>
                            <Route path="/" exact>
                                <Redirect to="/chat" />
                            </Route>
                            <Route
                                path="/chat"
                                render={() => {
                                    return (
                                        <ChatWindow
                                            parentHeight={height}
                                            isPortrait={isPortrait}
                                            maxHeight={73}
                                        />
                                    );
                                }}
                            />
                            <Route
                                path="/settings"
                                render={() => {
                                    return (
                                        <UserSettings
                                            parentHeight={height}
                                            isPortrait={isPortrait}
                                        />
                                    );
                                }}
                            />
                        </Switch>
                    </Box>
                </Box>
            </Router>
        </ThemeProvider>
    );
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
