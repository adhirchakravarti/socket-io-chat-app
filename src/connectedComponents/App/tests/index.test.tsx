import React from "react";
import {render} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {ThemeProvider} from "@material-ui/core/styles";
import {StylesProvider} from "@material-ui/styles/";
import CssBaseline from "@material-ui/core/CssBaseline";

import GlobalResponsiveFont from "../../../globalResponsiveFont";
import {lightTheme} from "../../../theme";
import store from "../../../store/store";

import App from "../index";

function getPreparedPage(store, theme): JSX.Element {
    const generateClassName = (rule, styleSheet) =>
        `${styleSheet.options.classNamePrefix}-${rule.key}`;
    return (
        <Provider store={store}>
            <StylesProvider generateClassName={generateClassName}>
                <ThemeProvider theme={theme}>
                    <CssBaseline />
                    <GlobalResponsiveFont />
                    <MemoryRouter>
                        <App />
                    </MemoryRouter>
                </ThemeProvider>
            </StylesProvider>
        </Provider>
    );
}

describe("<App />", () => {
    it("renders correctly and matches snapshot", () => {
        const {asFragment} = render(getPreparedPage(store, lightTheme));
        const initialRender = asFragment();
        expect(initialRender).toMatchSnapshot();
    });
    it("does not log errors in the console", () => {
        const spy = jest.spyOn(global.console, "error");
        render(getPreparedPage(store, lightTheme));
        expect(spy).not.toHaveBeenCalled();
    });
});
