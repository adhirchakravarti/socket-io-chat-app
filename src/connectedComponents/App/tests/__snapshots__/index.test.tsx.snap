// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root makeStyles-gridContainer makeStyles-gridContainer"
  >
    <div
      class="MuiBox-root MuiBox-root"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6"
          >
            Chat-App
          </h6>
        </div>
      </header>
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer MuiTabs-centered"
              role="tablist"
            >
              <a
                aria-disabled="false"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-selected MuiTab-labelIcon"
                role="tab"
                tabindex="0"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  <span
                    class="MuiBadge-root"
                    style="opacity: 1; webkit-transition: opacity 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 500ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
                      />
                    </svg>
                    <span
                      class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorSecondary MuiBadge-invisible"
                    >
                      0
                    </span>
                  </span>
                  Chat
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
              <a
                aria-disabled="false"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-labelIcon"
                role="tab"
                tabindex="-1"
              >
                <span
                  class="MuiTab-wrapper"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                      transform="scale(1.2, 1.2)"
                    />
                  </svg>
                  Settings
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
            <span
              class="PrivateTabIndicator-root PrivateTabIndicator-colorPrimary MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root makeStyles-content"
    >
      <div
        class="MuiPaper-root makeStyles-chatContainer makeStyles-chatContainer MuiPaper-elevation3 MuiPaper-rounded"
      >
        <div
          class="MuiBox-root MuiBox-root makeStyles-listContainer makeStyles-listContainer"
        >
          <ul
            class="MuiList-root makeStyles-list MuiList-padding"
          />
        </div>
        <div
          class="MuiBox-root MuiBox-root makeStyles-input"
        >
          <div
            class="MuiBox-root MuiBox-root makeStyles-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root makeStyles-input MuiFormControl-marginDense"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined"
                data-shrink="true"
              >
                Chat
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                  placeholder="Enter chat message and click the send button"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled PrivateNotchedOutline-legendNotched"
                  >
                    <span>
                      Chat
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Send
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
